### PR TITLE
8307176: Monkey Tester Application Part 2

### PR DESCRIPTION
**Pull request by andy-goryachev-oracle**
Adding changes to the MonkeyTester application accumulated since the last test sprint, from a separate repository
https://github.com/andy-goryachev-oracle/MonkeyTest

User preferences location:
The applications stores its user preferences (window location, etc.) in ~/.MonkeyTester directory.
- To use a different directory, redefine the "user.home" system property, -Duser.home=<...>.
- To disable saving, specify -Ddisable.settings=true vm agrument.

- ✔ replace setId() with some other way of setting component name
- ✔ change property for ui.settings dir
- ✔ add pie chart
- ✔ ComboBox: The current two buttons don't seem all that useful. I'm not even sure what, exactly, they do. What would be useful is a way to select the number of items in the list (like there is with ChoiceBox)
- ✔ ListView: Changing the selection model or checking / unchecking the "null focus model" option clears the list, which is unexpected. Given that there is a separate "clear list" button, it doesn't seem needed either.
- ✔ TextField: The default alignment of BASELINE_RIGHT is unexpected (unless there is a good reason, defaults for properties should match the API default to avoid surprises).
- ✔ Menu item: Window --> Open Modal Window "Platform.exit()" should not be the default choice (if you press <ENTER> the app will exit)
- ✔ TableView: The column sorting feature would be more useful if it were possible to have different data for each row
- ✔ Add Skin -> (null skin, set new skin) menus to all related controls to enable leak tests
- ✔ TableView: cell factory, cell value factory 
- ✔ TreeTableView: cell factory
- ✔ TreeView: cell factory
- ✔ add massive CJK text for JDK-8090110 / JDK-8089418
- ✔ clipboard monitor tool
- ✔ add to the status bar: JVM version, JFX version, current directory, screen scale
- ✔ WebView page
- ✔ Tools -> Keyboard Events Viewer
- ✔ System Info tool that reports all the details about the environment such as reported OS version, number of displays, env, system properties

To be continued... JDK-8316372

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8307176](https://bugs.openjdk.org/browse/JDK-8307176): Monkey Tester Application Part 2 (**Enhancement** - P4)

### Reviewers
 * [Ajit Ghaisas](https://openjdk.org/census#aghaisas) (@aghaisas - **Reviewer**)
 * [Karthik P K](https://openjdk.org/census#kpk) (@karthikpandelu - Committer) 🔄 Re-review required (review applies to [5ba98186](https://git.openjdk.org/jfx/pull/1241/files/5ba9818646184a71dc33b341943bdbf76db5d710))

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1241/head:pull/1241` \
`$ git checkout pull/1241`

Update a local copy of the PR: \
`$ git checkout pull/1241` \
`$ git pull https://git.openjdk.org/jfx.git pull/1241/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1241`

View PR using the GUI difftool: \
`$ git pr show -t 1241`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1241.diff">https://git.openjdk.org/jfx/pull/1241.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1241#issuecomment-1721582012)

### CODE REVIEW

**Comment by kevinrushforth** [comment]
Reviewers: @karthikpandelu @aghaisas

File: tests/manual/monkey/src/com/oracle/tools/fx/monkey/tools/KeyboardEventViewer.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import javafx.scene.control.TextArea;
+import javafx.scene.control.ToolBar;
+import javafx.scene.input.KeyEvent;
+import javafx.scene.layout.BorderPane;
+import com.oracle.tools.fx.monkey.util.FX;
+
+/**
+ * Keyboard Event Viewer
+ */
+public class KeyboardEventViewer extends BorderPane {
```
Comment (beldenfox) [inline tests/manual/monkey/src/com/oracle/tools/fx/monkey/tools/KeyboardEventViewer.java:38]: Yes! Thank you for adding a keyboard event viewer. It is long overdue.

On Mac and Linux dead keys are handled using InputMethodEvents so I find it essential to catch them and dump out some minimal information. Here's the code I use (`addToLog` simply appends the text to a TextArea with a newline at the end):

```
    private void inputMethodTextChangedEvent(InputMethodEvent e) {
        if (!e.getCommitted().isEmpty()) {
            addToLog("  Commit: " + e.getCommitted());
        }
        if (!e.getComposed().isEmpty()) {
            String composedText = "";
            for (InputMethodTextRun run : e.getComposed()) {
                composedText += run.getText();
            }
            addToLog("Composed: " + composedText);
        }
    }

```

Comment (andy-goryachev-oracle) [reply tests/manual/monkey/src/com/oracle/tools/fx/monkey/tools/KeyboardEventViewer.java:39]: Thank you so much - this is a very good suggestion.  Added.

**Review by aghaisas — APPROVED** [review]
Basic sanity testing looks good!

**Review by karthikpandelu — APPROVED** [review]
I have done basic testing and it looks good.
I do have some comments:
ComboBox:
- Option to enter number of items in the ComboBox is not yet added hence this point can be added in the follow-up task.
- I did notice that the clicking "Change Visible Row Count" button changes the Visible Row Count to 20 and then changes to 2 without clicking. And it makes the window unresponsive as well.

TableView: Adding column using "Add Column" adds the column with random number as header. Was this intentional?

Window -> Open Modal Window: Mouse click on "Platform.exit()” works as expected. However using keyboard (Tab and then ENTER) is not working.

**Comment by andy-goryachev-oracle** [comment]
Thank you all for reviewing!
I just wanted to say that the tools is still work in progress, but I think there is some value in releasing it in its current state.

Responding to Karthik's questions:
1. ComboBox page was quickly done for testing one particular bug.  I am planning to change it later with a more complete set of configuration options similar to other pages (btw, what is the scenario where the window becomes unresponsive?  is that another issue?  which platform are you running on?)
2. TableView: yes, it's intentional.
3. Platform.exit() button works but you must press SPACE not ENTER (unexpected, but that's how buttons seem to work in FX)

**Pull request closed without merging**